Update "ssl-config" to support X-Pack features

tvernum opened the pull request:
This commit upgrades the existing SSPL licensed "ssl-config" library
to include additional features that are supported by the X-Pack SSL
library.

This commit does not make any changes to X-Pack to use these new
features - it introduces them in preparation for their future use by
X-Pack.

The reindex module is updated to reflect API changes in ssl-config

tvernum committed:
Update "ssl-config" to support X-Pack features

This commit upgrades the existing SSPL licensed "ssl-config" library
to include additional features that are supported by the X-Pack SSL
library.

This commit does not make any changes to X-Pack to use these new
features - it introduces them in preparation for their future use by
X-Pack.

The reindex module is updated to reflect API changes in ssl-config

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

tvernum commented:
I've labelled this as `v8.0.0` only, because it's not needed in 7.x (it's only here to support migrating x-pack SSL onto this library).
We can talk about backporting it to 7.x if someone feels strongly, but I didn't make the changes with perfect BWC in mind, I'd need to do a review first.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslConfigurationLoader.java:
```diff
@@ -281,21 +287,22 @@ public SslConfiguration load(Path basePath) {
         }
-        return new SslConfiguration(trustConfig, keyConfig, verificationMode, clientAuth, ciphers, protocols);
+        final boolean isExplicitlyConfigured = hasSettings(settingPrefix);
+        return new SslConfiguration(isExplicitlyConfigured, trustConfig, keyConfig, verificationMode, clientAuth, ciphers, protocols);
```
`isExplicitlyConfigured` is used in watcher to know whether an SSL configuration really exists. It always "exists" - the SSL service always creates a context, but watcher wants to know whether the admin actually configured it, or whether the SSL service created a default one.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslConfigurationLoader.java:
```diff
@@ -281,21 +287,22 @@ public SslConfiguration load(Path basePath) {
-        final Path trustStorePath = resolveSetting(TRUSTSTORE_PATH, basePath::resolve, null);
+    protected SslTrustConfig buildTrustConfig(Path basePath, SslVerificationMode verificationMode, SslKeyConfig keyConfig) {
+        final List<String> certificateAuthorities = resolveListSetting(CERTIFICATE_AUTHORITIES, Function.identity(), null);
+        final String trustStorePath = resolveSetting(TRUSTSTORE_PATH, Function.identity(), null);
```
There's a number of changes that switch from `Path` to `String`. That's needed for the SSL certificates API. It should report the "path" as configured in the settings, not the path as on disk.

That is it reports `http.p12` not `${ES_HOME}/config/http.p12`.
So, all of these config classes needed to change to use `String` instead of `Path` and resolve the file at the last possible moment.

I considered creating a new class just for this (and still can) but on balance is seemed like passing around a new `ConfigPath` object would be more confusing.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/PemUtils.java:
```diff
@@ -73,6 +72,29 @@ private PemUtils() {
+     * @param passwordSupplier A password supplier for the potentially encrypted (password protected) key
+     * @return a private key from the contents of the file
+     */
+    public static PrivateKey readPrivateKey(Path path, Supplier<char[]> passwordSupplier) throws IOException, GeneralSecurityException {
```
This is just a new wrapper method that does the special exception handling we need (so that admins get useful error messages).

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/PemUtils.java:
```diff
@@ -109,13 +131,9 @@ public static PrivateKey readPrivateKey(Path keyPath, Supplier<char[]> passwordS
-        } catch (FileNotFoundException | NoSuchFileException e) {
-            throw new SslConfigException("private key file [" + keyPath.toAbsolutePath() + "] does not exist", e);
-        } catch (IOException | GeneralSecurityException e) {
-            throw new SslConfigException("private key file [" + keyPath.toAbsolutePath() + "] cannot be parsed", e);
```
Dumped this exception handling so that the wrapper method above can do a better, more consistent job of it.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/Pkcs11KeyConfig.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * A {@link SslKeyConfig} that builds a Key Manager from a keystore file.
+ */
+public class Pkcs11KeyConfig extends SslKeystoreConfig {
```
Not sure whether we want this - we don't have tests for it, and we talked about dropping PKCS#11 support.
But for now it's here since X-Pack has it.

tvernum replied:
I'm going to drop PKCS#11 support from this PR, and open a new PR to remove PKCS#11 support from X-Pack (in 8.0 only).
That way we have an explicit PR with a breaking change, and these PRs can remain non-issue PRs.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslConfiguration.java:
```diff
@@ -56,15 +57,18 @@
+    public SslConfiguration(boolean explicitlyConfigured, SslTrustConfig trustConfig, SslKeyConfig keyConfig,
+                            SslVerificationMode verificationMode, SslClientAuthenticationMode clientAuth,
+                            List<String> ciphers, List<String> supportedProtocols) {
+        this.explicitlyConfigured = explicitlyConfigured;
```
See explanation in SslConfigurationLoader

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/PemKeyConfig.java:
```diff
@@ -8,84 +8,148 @@
+     * @param keyPassword    Password for the private key (or empty is the key is not encrypted)
+     * @param configBasePath The base directory from which config files should be read (used for diagnostic exceptions)
+     */
+    public PemKeyConfig(String certificate, String key, char[] keyPassword, Path configBasePath) {
```
The change from `Path` to `String` is explained in `SslConfigurationLoader`

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/PemKeyConfig.java:
```diff
@@ -8,84 +8,148 @@
+            throw SslFileUtil.ioException(KEY_FILE_TYPE, List.of(path), e);
         } catch (GeneralSecurityException e) {
-            throw new SslConfigException("cannot load ssl private key file [" + key.toAbsolutePath() + "]", e);
+            throw SslFileUtil.securityException(KEY_FILE_TYPE, List.of(path), e);
```
Lots of exception handling has been unified. See comments in `SslFileUtil`

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslFileUtil.java:
```diff
@@ -0,0 +1,126 @@
+/**
+ * Utility methods for common file handling in SSL configuration
+ */
+final class SslFileUtil {
```
This is a new class, mostly to handle generating exceptions for particular file related problems.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslFileUtil.java:
```diff
@@ -0,0 +1,126 @@
+        return ioException(fileType, paths, cause, null);
+    }
+
+    static SslConfigException ioException(String fileType, List<Path> paths, IOException cause, String detail) {
```
One of the big things we did to try and make X-Pack SSL setup easier was write clear & explicit error messages for all the common problems that people run into:
- File not found
- File permissions wrong
- File in wrong directory
- Missing password
- Wrong password

This class replicates those messages into ssl-config and centralises them.
Some of the messages already existed in this library, but the X-Pack ones are encountered more frequently, and have been better turned for readability.

That said, I did try and improve some of them along the way.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslKeystoreConfig.java:
```diff
@@ -0,0 +1,200 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public abstract class SslKeystoreConfig implements SslKeyConfig {
```
This is a base class for PKCS#11 and File-Keystore (JKS, PKCS#12) support.
The X-Pack model of having 1 class that handled both was really fragile - we often wrote code that assume that the `file` was non-null.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/StoredCertificate.java:
```diff
@@ -0,0 +1,80 @@
+ * Information about a certificate that is locally stored.It includes a reference to the {@link X509Certificate} itself,
+ * as well as information about where it was loaded from.
+ */
+public class StoredCertificate {
```
This is used for the `_ssl/certificates` API.
That API uses a `CertificateInfo` which depends on `ToXContent` etc, which isn't available inside this library. So this class is a lightweight representation of the information needed for that API, and X-Pack will convert from this class to the X-Pack class.

tvernum commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/merge-ssl-config/common-ssl-base

BigPandaToo reviewed (approved):
I added a couple of nit comments.
LGTM otherwise

tvernum committed:
Merge branch 'master' into feature/merge-ssl-config/common-ssl-base

tvernum committed:
Improve toString on SslKeystoreConfig

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslKeystoreConfig.java:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    public boolean hasKeyPassword() {
+        return Arrays.equals(storePassword, keyPassword) == false;
```
Yeah, that's kind of weird.
It was only used in the `toString`, so I removed it.

These classes track the keystore and key password separately, but the key password default to the keystore password (which is normal for keystores).
So, the intent here was "do we have a key password that has been set to something other than the keystore password?" but we can do a better job of that, which is what I did in the updated toString.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/KeyStoreUtil.java:
```diff
@@ -91,9 +91,10 @@ static KeyStore buildKeyStore(Collection<Certificate> certificateChain, PrivateK
 
     /**
      * Construct an in-memory keystore with multiple trusted cert entries.
+     *
```
I wouldn't say so.
Everything in that file has blank line between the method description & the parameters, so this format is consistent with the rest of the code.

jkakavas commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslKeyConfig.java:
```diff
@@ -31,5 +36,27 @@
+    /**
+     * @return A collection of {@link X509Certificate certificates} used by this config.
+     */
+    Collection<? extends StoredCertificate> getConfiguredCertificates();
```
Do we foresee exteding StoredCertificate ? Can we make the return type non generic for now ?

jkakavas commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslKeyConfig.java:
```diff
@@ -31,5 +36,27 @@
+    List<Tuple<PrivateKey, X509Certificate>> getKeys();
+
+    /**
+     * @return A collection of {@link X509Certificate certificates} used by this config.
```
nit: update javadoc with corret return type

jkakavas commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/PemKeyConfig.java:
```diff
@@ -8,84 +8,148 @@
+        if (certificates.isEmpty()) {
+            return List.of();
+        }
+        final Certificate certificate = certificates.get(0);
```
nit: rename certificate here for readability ( vs this.certificate)

jkakavas commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslFileUtil.java:
```diff
@@ -0,0 +1,126 @@
+
+        if (cause.getCause() instanceof UnrecoverableKeyException) {
+            message += " - this is usually caused by an incorrect password";
+        } else if (cause != null && cause.getMessage() != null) {
```
cause can't be null here, otherwise we'd have thrown an NPE in cause.getCause() above. We can check that cause is not null above and not go into this conditional at all

jkakavas reviewed (commented):
Thanks for taking the time to add comments for things that would most probably raise questions, that was very helpful. Added some suggestions and comments, mostly nits

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/PemKeyConfig.java:
```diff
@@ -8,84 +8,148 @@
+        boolean first = true;
+        for (Certificate cert : certificates) {
+            if (cert instanceof X509Certificate) {
+                info.add(new StoredCertificate((X509Certificate) cert, this.certificate, "PEM", null, first));
```
Probably not - but that's what X-Pack currently does.

It _should_ be - if the certificate path points to a PEM file with multiple certificates, then they _should_ be a chain ordered from leaf to issuer to root - but there's no guarantee that it's true.

However, other things will break if it's not. When we build a keystore we will assume the certificates that we read form a valid chain, and `Keystore.setKeyEntry` will throw an exception if they're not.

tvernum replied:
I added a test to show that `createKeyManager` fails if the chain is in the wrong order. I think that is sufficient for the assumption in this method.

We do have a "private key" for the leaf (first element) certificate. It might be the wrong key (none of the config classes check that in this method), but we have one.
And the core functionality of the config (building a key manager) will fail.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/KeyStoreUtil.java:
```diff
@@ -148,5 +159,90 @@ static X509ExtendedTrustManager createTrustManager(@Nullable KeyStore trustStore
+    static Stream<KeyStoreEntry> stream(KeyStore keyStore,
+                                        Function<GeneralSecurityException, ? extends RuntimeException> exceptionHandler) {
+        try {
+            return Collections.list(keyStore.aliases()).stream().map(a -> new KeyStoreEntry(keyStore, a, exceptionHandler));
```
Yes, because `Collections.list` forms a list from an `Enumeration<?>` which is the only useful way (I think) of turning an Enumeration (which is what Keystore returns) into a `Stream`

tvernum replied:
In Java 9, you can turn an `Enumeration` into an `Iterator` and then an `Iterator` into a `Spliterator` and then a `Spliterator` into a `Stream`. But that's a lot of conversions just to avoid forming a list with a handful of elements in it.

jkakavas replied:
thanks

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/KeyStoreUtil.java:
```diff
@@ -148,5 +159,90 @@ static X509ExtendedTrustManager createTrustManager(@Nullable KeyStore trustStore
+            try {
+                final Certificate[] certificates = store.getCertificateChain(alias);
+                if (certificates == null || certificates.length == 0) {
+                    return List.of();
```
No, it is intentional that this returns a more useful API than `Keystore`.
I can document it if you'd like, but I feel pretty strongly that the purpose writing this wrapper is to avoid those cumbersome low level details from `Keystore`.

jkakavas replied:
No need to document additionally I think, reasoning makes sense to me

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/StoreKeyConfig.java:
```diff
@@ -9,87 +9,84 @@
+
+    @Override
+    public SslTrustConfig asTrustConfig() {
+        final String trustStoreAlgorithm = TrustManagerFactory.getDefaultAlgorithm();
```
I think there's value in being clear about which layer deals with "this is the most common value and I will apply it as a default" vs the layers that say "it is your job to know what you want". 
I don't know if we have the right answer here, but it's a problem to go too far the other way and find that classes that ought to be simple end up with a dozen constructors and utility methods that make them more complex.

tvernum committed:
Address Feedback

- Remove PKCS#11
- Improve exception messages
- Simplify return types

tvernum committed:
Merge branch 'master' into feature/merge-ssl-config/common-ssl-base

tvernum committed:
Add test for unordered PEM cert chain

PemKeyConfig.getConfiguredCertificates relies on the certificate file
containing a chain from leaf -> issuer -> .. -> root

This commit adds a test to ensure that createKeyManager will fail if
that order is not correct, and therefore the assumption in
getConfiguredCertificates is acceptable

tvernum committed:
Add javadoc on KeyStoreEntry

jkakavas reviewed (approved):
LGTM
